pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,257 @@
+		return ctrl.Result{}, nil
+	}
+
+	var kubeconfig v1alpha1.ClusterKubeconfig
+
+	if err := r.Get(ctx, req.NamespacedName, &kubeconfig); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return ctrl.Result{}, err
+		}
+
+		l.Info("kubeconfig not found, will be created")
+		kubeconfig.Name = cluster.Name
+		kubeconfig.Namespace = cluster.Namespace
+		kubeconfig.Spec.Kubeconfig = v1alpha1.ClusterKubeconfigData{
+			Kind:       "Config",
+			APIVersion: "v1",
+		}
+		kubeconfig.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: cluster.APIVersion,
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			},
+		}
+		kubeconfig.Spec.Kubeconfig.Contexts = []v1alpha1.ClusterKubeconfigContextItem{
+			{
+				Name: cluster.Name,
+				Context: v1alpha1.ClusterKubeconfigContext{
+					Cluster:   cluster.Name,
+					AuthInfo:  "oidc@" + cluster.Name,
+					Namespace: "default",
+				},
+			}}
+		kubeconfig.Spec.Kubeconfig.CurrentContext = cluster.Name
+		kubeconfig.Status.Conditions.SetConditions(v1alpha1.TrueCondition(v1alpha1.KubeconfigCreatedCondition, "NewCreation", ""))
+	}
+
+	// get oidc info from organization
+	oidc, err := r.getOIDCInfo(ctx, cluster.Namespace)
+	if err != nil {
+		l.Error(err, "skip reconcile, oidc data not fetched")
+		return ctrl.Result{}, nil
+	}
+
+	// get cluster connection data from cluster secret
+	var secret corev1.Secret
+	err = r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, &secret)
+	if err != nil {
+		l.Error(err, "skip reconcile, secret data not fetched")
+		return ctrl.Result{}, nil
+	}
+
+	// collect cluster connection data and update kubeconfig
+	rootKubeCfg := secret.Data["kubeconfig"]
+	kubeCfg, err := clientcmd.Load(rootKubeCfg)
+	if err != nil {
+		l.Error(err, "unable to load kubeconfig")
+		return ctrl.Result{}, err
+	}
+
+	var clusterCfg *clientcmdapi.Cluster
+	for _, v := range kubeCfg.Clusters {
+		clusterCfg = v
+		break
+	}
+
+	// collect oidc data and update kubeconfig
+	clientutil.CreateOrPatch(ctx, r.Client, &kubeconfig, func() error {
+		kubeconfig.Spec.Kubeconfig.Clusters = []v1alpha1.ClusterKubeconfigClusterItem{
+			{
+				Name: cluster.Name,
+				Cluster: v1alpha1.ClusterKubeconfigCluster{
+					Server:                   clusterCfg.Server,
+					CertificateAuthorityData: clusterCfg.CertificateAuthorityData,
+				},
+			}}
+		kubeconfig.Spec.Kubeconfig.AuthInfo = []v1alpha1.ClusterKubeconfigAuthInfoItem{
+			{
+				Name: "oidc@" + cluster.Name,
+				AuthInfo: v1alpha1.ClusterKubeconfigAuthInfo{
+					AuthProvider: clientcmdapi.AuthProviderConfig{
+						Name: "oidc",
+						Config: map[string]string{
+							"client-id":      oidc.ClientID,
+							"client-secret":  oidc.ClientSecret,
+							"idp-issuer-url": oidc.IssuerURL,
+						},
+					},
+				},
+			},
+		}
+		kubeconfig.Status.Conditions.SetConditions(v1alpha1.TrueCondition(v1alpha1.KubeconfigReadyCondition, "Complete", ""))
```

Comment:
Not sure if this is persisted without `clientutil.PatchStatus` as Status is a Subresource? Have you tested this?
But if you refactor the Status Reporting, this might move away anyhow.

Reply:
Patch status is refactored by 16b7fb7af545499f90a2f168752ee2ceb335a60a and also tested as part of unit tests